samza-core/src/test/java/org/apache/samza/clustermanager/MockContainerAllocator.java

```diff
@@ -23,16 +23,29 @@
 import java.lang.reflect.Field;
 
 import java.util.Map;
+import java.util.concurrent.CountDownLatch;
 
 public class MockContainerAllocator extends ContainerAllocator {
   public int requestedContainers = 0;
+  private CountDownLatch latch = new CountDownLatch(1);
 
   public MockContainerAllocator(ClusterResourceManager manager,
                                 Config config,
                                 SamzaApplicationState state) {
     super(manager, config, state);
   }
 
+  /**
+   * Causes the current thread to block until the provided number of containers have started.
+   *
+   * @param numExpectedContainers  the number of containers expected to start
+   * @throws InterruptedException  if the current thread is interrupted while waiting
+   */
+  void awaitContainersStart(int numExpectedContainers) throws InterruptedException {
+    latch = new CountDownLatch(numExpectedContainers);
```

Comment:
I would suggest against an unbounded await. How about some reasonable timeout and then fail the test?

Instead of reassigning the count down latch you could use a semaphore and acquire the expected number of permits here and release them as containers start up.